converter/internal/common/weaveworks_server.go

```diff
@@ -0,0 +1,81 @@
+package common
+
+import (
+	"github.com/grafana/agent/converter/diag"
+	"github.com/weaveworks/common/server"
+)
+
+func DefaultWeaveWorksServerCfg() server.Config {
+	cfg := server.Config{}
+	// NOTE: due to a bug in promtail, the default server values for loki_push_api are not applied currently, so
+	// we need to comment out the following line.
+	//cfg.RegisterFlags(flag.NewFlagSet("", flag.PanicOnError))
+	return cfg
+}
+
+func ValidateWeaveWorksServerCfg(cfg server.Config) diag.Diagnostics {
```

Comment:
I'm commenting on this line just to get the conversation going.

I think we should discuss and narrow in a consistent approach is for adding diags during the conversion. It seems like we currently have too much variance in how we do things which makes it hard for me to follow along if I'm looking at the converter validations:
- a general `validate.go` in the *convert package
- various `validate*` funcs in flow component specific files, ie. `prometheusconvert.validatePrometheusScrape` which returns `diag.Diagnostics`
- adding them directly to the property on a builder, ie. `build.AppendKubernetesSDs`
- passing a `*diag.Diagnostics` into another func, ie. `build.convertStage`

Not to suggest this is the best final pattern especially since I am interested in the builder pattern but the current pattern I have in place for prom/static converters exists like this:
- Any diags for the main parsing is done in the top level `prometheusconvert.Convert` func
- validation diags aren't added while appending components (maybe this is too strict for all cases, but it does separate the responsibilities)
- `validation.go` provides a top level `validate` func that returns all validation diags not covered by `prometheusconvert.Convert` 
  - The validation is ordered by the order of the properties in the source config type
  - Any non-component specific validation exists in this file
  - Any component specific validation calls a validation func which exists in that component's specific convert file

Reply:
Yeah, I've noticed these differences too and I agree we could bring things to be a bit more consistent (this or preferably a future PR). 

However there are a few reasons why I am hesitant about a one big validate step followed by building the config separately:
* It spreads the code around a bit... I would prefer if adding a new feature support was ideally in one place. This is somewhat alleviated by the convention of a `validate*` function though in the same file as the conversion code.
* However, it is still possible to forget to call validation or to forget to write it.
* In one-step validation approach, some issues to validate would require iterating over arrays and layers of `nil` checking and reaching deep into the config struct, [for example this one](https://github.com/grafana/agent/blob/322ffd1262da92e9214cd673c5e3653c665bc3e3/converter/internal/promtailconvert/internal/build/journal.go#L22) is in `scrape_configs[*].journal.max_age`. Since we'll already be walking through this config later to convert it, I found it redundant to walk through it in order to validate it.

Here's an idea how I think we could make the validation and conversion more consistent:

We could enforce the interface for config fragments conversions and force via API always validating when converting - this way it is harder to forget to call validate. An example interface could be like this:
```
// Converts config from F to T after validating and providing diagnostics.
type ConfigConverter[F interface{}, T interface{}] interface {
	Convert(fCfg F) (T, diag.Diagnostics)
}
```

And an example implementation and using of it:
```
type KubernetesSDConverter struct {
}

func (k KubernetesSDConverter) Convert(fCfg promkubernetes.SDConfig) (kubernetes.Arguments, diag.Diagnostics) {
	// here goes the current implementation
	panic("implement me")
}

func actualConversionCode(fCfg promkubernetes.SDConfig, f *builder.File, diags *diag.Diagnostics) {
	args, newDiags := KubernetesSDConverter{}.Convert(fCfg)
	diags.AddAll(newDiags)
	f.Body().AppendBlock(common.NewBlockWithOverride([]string{"discovery", "kubernetes"}, "default", args))
}
```

Alternatively I can refactor my code to match the pattern we have in the other converters - it may end up being a bit more verbose, but it'd be a feasible option too. 

Lastly we can leave the mixed approach - I agree being consistent is better, but calling validation is also not the most complex part around here, so I think it is not too much of an issue.

Reply:
for future us: we talked offline about moving forward as is for now and thinking about a refactor on some of this stuff after things settle. both approaches are very functional, and we can easily test any refactor.